contribute/design/1.x/ndm/20190705-ndm-disk-identification.md

```diff
@@ -0,0 +1,214 @@
+from node to node and even attached to different ports on the same node. The unique
+identification solution will help the consumers of NDM to track the disk
+and thus the data movement within the cluster.
+
+### Goals
+
+- A unique disk identification mechanism that will work across the cluster on
+  disks having atleast a GPT label
+- Should be able to identify same disk attached at multiple paths.
+- Should be able to generate separate IDs for partitions
+
+### Non-Goals
+
+- Identification of disk attached to 2 nodes at the same time
+
+## Proposal
+
+### User Stories
+
+#### Identify a physical disk
+NDM should be able to generate a unqiue id for a physical disk attached to the node
+and this id should not change even if the disk is attached at a different port on the
+same host or moved to a different node or altogether moved to a different cluster.
+
+#### Identify a virtual disk
+NDM should be able to generate a unique id for Virtual Disks(which may not have all 
+relevant information in VPD). This id should also persist between restarts of the node,
+attaching to a different machine or a different SCSI port of the same machine.
+
+### Implementation Details/Notes/Constraints
+
+The implementation details for the disk identification. The current implementation 
+works well for physical disks which have a unique WWN and model assigned by the 
+manufacturer. But the implementation fails for `Virtual_disk`, because the fields
+are provider dependent and are not guaranteed to be unique. 
+
+#### Current Implementation
+The following data is fetched from the disk: 
+1. `ID_WWN`
+2. `ID_MODEL`
+3. `ID_SERIAL_SHORT`
+4. `ID_VENDOR`
+5. `ID_TYPE`
+
+If the `ID_MODEL` is not `Virtual_disk` or `EphemeralDisk`:
+	- field 1-4 are appended
+	- md5 hash of the result is generated which will be UID for the disk by NDM
+
+If the `ID_TYPE` is empty or `ID_MODEL` is `Virtual_disk` or `EphemeralDisk`:
+	- There is a chance that the values in fields 1-4 will be either empty or all 
+	  the same depending on the provider.
+	- fields 1-4 are appended. Along with it, the hostname and DEVNAME(`/dev/sda`) is
+	  appended to the result
+	- md5 hash of the result is generated which will be UID for the disk by NDM
+
+#### Shortcomings of current implementation
+In case of virtual disks, since we are using the DEVNAME for generating UID, everytime 
+the same disk comes at a different path, the UID gets changed. This results in getting 
+the device identified as a new disk. This will lead to data corruption as the disk that 
+was being claimed by some users has now came up with a new UID, but with the old data.
+Also, since this disk is now in an Unclaimed state, it will be available for others to
+claim possibly leading to data loss
+
+#### Proposed Implementation
+
+1. We try to generate the UUID using the current implementation. 
+
+2. If the `ID_TYPE` is empty or `ID_MODEL` is `Virtual_disk` or `EphemeralDisk` and `ID_WWN` is empty 
```

Comment:
If we go the ID_MODEL identification route, we may add also `QEMU_HARDDISK`, which is presented by KVM-based hosts.

```suggestion
2. If the `ID_TYPE` is empty or `ID_MODEL` is `Virtual_disk` or `EphemeralDisk` or `QEMU_HARDDISK` and `ID_WWN` is empty 
```

Reply:
For `QEMU_HARDDISK`, we will follow the same path as of `Virtual_Disk`, if we are unable to get a WWN